lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1126,7 +1133,46 @@ TypeExpr *PreCheckExpression::simplifyTypeExpr(Expr *E) {
                                         ResultTypeRepr);
     return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
   }
-
+  
+  // Fold 'P & Q' into a composition type
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(E))
```

Comment:
Here I go beyond the scope of the proposal and form a `TypeExpr` from `(P & Q).self` which was not possible with the previous syntax; I considered it a bug but wanted to make sure this change is wanted.